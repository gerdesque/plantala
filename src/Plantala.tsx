import React from 'react';
import Header from './header/Header';
import './Plantala.css';

function Plantala() {
  return (
    <div className="plantala">
      <Header />
    </div>
  );
}

export default Plantala;
