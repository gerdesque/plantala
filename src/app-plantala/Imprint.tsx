import React from 'react';
import { Dialog, DialogContent, DialogContentText, DialogTitle, Link } from '@material-ui/core';

interface IImprintProps {
  open: boolean
  onClose: any
}

export default function Imprint({ open, onClose }: IImprintProps) {

  return (
    <Dialog
        open={open}
        onClose={onClose}
        aria-labelledby='responsive-dialog-title'
      >
        <DialogTitle id='responsive-dialog-title'>{'Malen. Strahlen. Plantala!'}</DialogTitle>
        <DialogContent>
          <DialogContentText>
            Plantala ist eine Webapplikation, bei der du aus ästhetischen Zell- und Pflanzenstrukturen dein eigenes Mandala erstellst. Danach kannst du es teilen, ausdrucken und ausmalen. Plantala verknüpft Informationen über die verfügbaren Pflanzen mit digitaler Kreativität und analoger Entspannung. Damit ist sie die optimale Anti-Burnout-Anwendung für corona-gestresste Großstädter, die auch beim Gedankenloslassen noch etwas dazulernen wollen.
          </DialogContentText>
          <DialogContentText>
            Die Website Plantala wurde von Anne Mühlich und Gerd Müller im Rahmen des Kulturhackathons Coding da Vinci Niedersachsen 2020/2021 entwickelt. 
          </DialogContentText>
          <DialogContentText>
            Der Quellcode läuft unter einer MIT Lizenz. Die von der Universität Göttingen bereitgestellten und bearbeiteten Digitalisate von Botanischen Lehrtafeln wurden unter einer CC BY-SA 4.0 Lizenz veröffentlicht und sind <Link href="https://sammlungen.uni-goettingen.de/sammlung/slg_1053/" target="_blank" rel="noreferrer" > {'hier'} </Link> zu finden.
          </DialogContentText>
          <DialogContentText>
            Der Hintergroundsound unter dem Titel „AUTUMN SUNSET“ stammt von Jason Shaw und läuft unter einer CC-BY-3.0 United States Lizenz auf <Link href="https://freemusicarchive.org/music/Jason_Shaw/Audionautix_Acoustic/AUTUMN_SUNSET_________________1-36" target="_blank" rel="noreferrer" > {'freemusicarchive.org'} </Link>.
          </DialogContentText>
        </DialogContent>
      </Dialog>
  );
}
