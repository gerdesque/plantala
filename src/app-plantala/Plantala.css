.plantala {
  height: 100vh;
  width: 100vw;
  display: grid;
  grid-template-rows: auto 1fr auto;
}

@media print {
  html, body { width: 210mm; height: 297mm;}
  footer, img, button, h5 { display: none !important; }
  .MuiAppBar-colorPrimary {box-shadow: none !important;}
  h2::before { content: "Malen. Strahlen. "; }
  h2::after { content: "!"; }
  .MuiContainer-root { max-width: 1600px !important; padding: 0 !important }
}