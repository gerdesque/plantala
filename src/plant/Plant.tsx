import React from 'react';
import { Typography } from '@material-ui/core';

export interface IPlant {
  active: boolean
  name: string
  source: string
  amount?: number
  description?: string
  order ?: number
  rotation?: number
  scale?: number
  size?: number
}

export default function Plant({name, source, description}: IPlant) {

  return (
    <div className="plant">
      <Typography variant="h2">
        {name}
      </Typography>
    </div>
  )
}
