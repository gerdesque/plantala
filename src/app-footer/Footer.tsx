import React from 'react';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import { Avatar, Grid } from '@material-ui/core';
import AvatarGroup from '@material-ui/lab/AvatarGroup';
import { IPlant } from '../plant/Plant';
import Drawer from '../drawer/Drawer';
import { avatarCount } from '../utils/Utils';

interface IFooterProps {
  selectedPlants: IPlant[],
  activePlant: IPlant,
  setActivePlant: any,
  transformPlant: any
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    footer: {
      padding: theme.spacing(2),
      backgroundColor: theme.palette.primary.main,
    },
    avatar: {
      width: theme.spacing(8),
      height: theme.spacing(8),
      borderColor: theme.palette.secondary.main,
      cursor: 'pointer'
    }
  }),
);

export default function Footer({selectedPlants, activePlant, setActivePlant, transformPlant}: IFooterProps) {
  const classes = useStyles();
  const avatars = selectedPlants.slice(0, avatarCount).map(plant =>
    <Avatar
      key={plant.name}
      alt={plant.name}
      src={`${process.env.PUBLIC_URL}/assets/${plant.source}`}
      className={classes.avatar}
      onClick={() => setActivePlant(plant)}
    />);

  return (
    <footer className={classes.footer}>
      <Grid container spacing={2} justify="center">
        <AvatarGroup max={avatarCount}>
          {avatars}
        </AvatarGroup>
        {activePlant && <Drawer
          activePlant={activePlant}
          setActivePlant={setActivePlant}
          transformPlant={transformPlant}
        />}
      </Grid>
    </footer>
  );
}
