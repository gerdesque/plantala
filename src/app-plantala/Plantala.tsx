import React, { Component } from 'react';
import Header from '../app-header/Header';
import Main from '../app-main/Main';
import Footer from '../app-footer/Footer';
import './Plantala.css';
import { plantItems } from '../plantItems';
import { IPlant } from '../plant/Plant';

export enum Action {
  Start = "Start",
  Select = "Select",
  Done = "Done",
  Again = "Again"
}

function* getNextAction(): IterableIterator<Action> {
  let current = Action.Start;
  while (true) {
    yield current;
    yield Action.Select;
    yield Action.Done;
    yield Action.Again;
  }
}

const nextAction = getNextAction();
nextAction.next();

interface IPlantalaState {
  plants: IPlant[],
  activePlant: IPlant,
  action: Action
}

class Plantala extends Component {
  state: IPlantalaState = {
    // values to be displayed in <Card />
    plants: plantItems,
    // active plant for <Plant />
    activePlant: {} as IPlant,
    // action mode to be displayed in <Main />
    action: Action.Start,
  }

  setAction = () => {
    this.setState({action: nextAction.next().value});
  }

  activatePlant = (selectedPlant:IPlant) => {
    const filteredArray =
      this.state.plants.map(plant => plant === selectedPlant ? { ...plant, active: !plant.active } : plant);
    this.setState({ plants: filteredArray })
  }

  render = () => {
    const { plants, action } = this.state;
    const selectedPlants = plants.filter((plant) => (plant.active === true));

    return (
      <div className="plantala">
        <Header />
        <Main
          plants={plants}
          selectedPlants={selectedPlants}
          action={action}
          setAction={this.setAction}
          setSelectedPlant={this.activatePlant} />
        <Footer
          selectedPlants={selectedPlants} />
      </div>
    );
  }
}

export default Plantala;
