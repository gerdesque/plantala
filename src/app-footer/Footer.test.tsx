import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import { Avatar } from '@material-ui/core';
import AvatarGroup from '@material-ui/lab/AvatarGroup';
import Footer from './Footer';
import { IPlant } from '../plant/Plant';
import { Action } from '../app-plantala/Plantala';

describe('Footer', () => {

  let shallow: any;
  let wrapper: any;

  beforeEach(() => {
    shallow = createShallow();
    wrapper = shallow(<Footer
      selectedPlants={[]}
      activePlant={{} as IPlant}
      setActivePlant={jest.fn()}
      transformPlant={jest.fn()}
      action={Action.Select}
      />);
  });

  it('should render correctly', () => expect(wrapper).toMatchSnapshot());

  it('should render a <footer /> container', () => {
    expect(wrapper.find('footer').length).toEqual(1);
  });

  it('should display an <AvatarGroup />', () => {
    expect(wrapper.find(AvatarGroup)).toHaveLength(1);
  });

  it('should display zero <Avatar />s', () => {
    expect(wrapper.find(Avatar)).toHaveLength(0);
  });

});

describe('Non-Empty Footer', () => {

  let shallow: any;
  let wrapper: any;

  beforeEach(() => {
    shallow = createShallow();
  });

  it('should display 2 <Avatar />s', () => {
    const selectedPlants = [
      {name: '1', source: '1.png', selected: true},
      {name: '2', source: '2.png', selected: true},
    ]
    wrapper = shallow(<Footer
      selectedPlants={selectedPlants}
      activePlant={{} as IPlant}
      setActivePlant={jest.fn()}
      transformPlant={jest.fn()} />);
    expect(wrapper.find(Avatar)).toHaveLength(2);
  });

  it('should trigger setActivePlant when <Avatar /> component is clicked', () => {
    const selectedPlants = [
      {name: '1', source: '1.png', selected: true},
      {name: '2', source: '2.png', selected: true},
    ]
    const setActivePlant = jest.fn();

    wrapper = shallow(<Footer
      selectedPlants={selectedPlants}
      activePlant={{} as IPlant}
      setActivePlant={setActivePlant}
      transformPlant={jest.fn()} />);

   wrapper.find(Avatar).first().simulate('click');
   expect(setActivePlant).toHaveBeenCalledTimes(1);
  });


  it('should display no more than 5 <Avatar />s', () => {
    const selectedPlants = [
      {name: '1', source: '1.png', selected: true},
      {name: '2', source: '2.png', selected: true},
      {name: '3', source: '3.png', selected: true},
      {name: '4', source: '4.png', selected: true},
      {name: '5', source: '5.png', selected: true},
      {name: '6', source: '6.png', selected: true},
    ]

    wrapper = shallow(<Footer
      selectedPlants={selectedPlants}
      activePlant={{} as IPlant}
      setActivePlant={jest.fn()}
      transformPlant={jest.fn()} />);
    expect(wrapper.find(Avatar)).toHaveLength(5);
  });

});