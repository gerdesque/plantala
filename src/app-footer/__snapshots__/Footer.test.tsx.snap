// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer should render correctly 1`] = `
<footer
  className="makeStyles-footer-1"
>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    container={true}
    direction="column"
    justify="center"
  >
    <WithStyles(ForwardRef(Grid))
      className="makeStyles-plantValues-3"
      container={true}
      justify="center"
      spacing={2}
    >
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        id="rotation-slider"
      >
        Rotation
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Slider))
        aria-labelledby="rotation-slider"
        defaultValue={0}
        max={360}
        onChange={[Function]}
        valueLabelDisplay="auto"
      />
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        id="rotation-slider"
      >
        Menge
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Slider))
        aria-labelledby="amount-slider"
        defaultValue={8}
        marks={true}
        max={20}
        min={4}
        onChange={[Function]}
        valueLabelDisplay="auto"
      />
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        id="rotation-slider"
      >
        Größe
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Slider))
        aria-labelledby="scale-slider"
        defaultValue={1}
        marks={true}
        max={2}
        min={0.1}
        onChange={[Function]}
        step={0.1}
        valueLabelDisplay="auto"
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      container={true}
      justify="center"
      spacing={2}
    >
      <WithStyles(ForwardRef(AvatarGroup))
        max={5}
      />
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</footer>
`;
